presto-main/src/main/java/io/prestosql/operator/CollectOperator.java

```diff
@@ -0,0 +1,151 @@
+/// The collected pages' value are used for creating a run-time filtering constraint (for probe-side table scan in an inner join).
+/// We support only small build-side pages (which should be the case when using "broadcast" join).
+public class CollectOperator
+        implements Operator
+{
+    public static class CollectOperatorFactory
+            implements OperatorFactory
+    {
+        private final int operatorId;
+        private final PlanNodeId planNodeId;
+        private final DynamicFilter dynamicFilter;
+
+        public CollectOperatorFactory(int operatorId, PlanNodeId planNodeId, DynamicFilter dynamicFilter)
+        {
+            this.operatorId = operatorId;
+            this.planNodeId = requireNonNull(planNodeId, "planNodeId is null");
+            this.dynamicFilter = dynamicFilter;
+        }
+
+        @Override
+        public Operator createOperator(DriverContext driverContext)
+        {
+            CompletableFuture<TupleDomain<ColumnHandle>> dynamicPredicateFuture = new CompletableFuture<>();
+            dynamicFilter.addFuture(dynamicPredicateFuture);
+            return new CollectOperator(
+                    driverContext.addOperatorContext(operatorId, planNodeId, CollectOperator.class.getSimpleName()),
+                    dynamicFilter, dynamicPredicateFuture);
+        }
+
+        @Override
+        public void noMoreOperators()
+        {
+            dynamicFilter.noMoreFutures();
+        }
+
+        @Override
+        public OperatorFactory duplicate()
+        {
+            throw new UnsupportedOperationException("duplicate() is not supported for CollectOperatorFactory");
+        }
+    }
+
+    private final OperatorContext context;
+    private boolean finished;
+    private Page current;
+    private final DynamicFilter.Collector predicateCollector;
+    private final CompletableFuture<TupleDomain<ColumnHandle>> dynamicPredicateFuture;
+    private long remainingRowCount;
+
+    public CollectOperator(OperatorContext context, DynamicFilter dynamicFilter, CompletableFuture<TupleDomain<ColumnHandle>> dynamicPredicateFuture)
+    {
+        this.context = requireNonNull(context, "context is null");
+        this.predicateCollector = dynamicFilter.collector();
+        this.dynamicPredicateFuture = dynamicPredicateFuture;
+        this.remainingRowCount = getDynamicFilteringMaxRowCount(context.getSession());
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return context;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return current == null && !finished;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        verify(!finished, "CollectOperator: addInput() may not be called after finish()");
+        current = page;
+        if (dynamicPredicateFuture.isCancelled()) {
+            return;  // the predicate became too large.
+        }
+        if (remainingRowCount >= page.getPositionCount()) {
```

Comment:
Apart from a limit on number of positions, in the older implementation we had also used a limit on size for the case where join column is a long VARCHAR
https://github.com/prestodb/presto/pull/9453#issuecomment-353018791

Reply:
Good idea, thanks!
Added in 4dd335f.